### ISSUE TITLE
Some iOS tests are failing to launch test app

### ISSUE THREAD

**Issue by safern** [dotnet/runtime#45181]
I just saw a couple of failures on a PR on iOS, with:

```
[16:59:08] dbug: 16:59:08.5015570 Xamarin.Hosting: Simulator watchdogs are already disabled for 'iOS 13.5 (17F61) - iPhone X'.
[16:59:08] dbug: 16:59:08.5075460 Xamarin.Hosting: Device 'iOS 13.5 (17F61) - iPhone X' is already booted and the simulator is running.
[16:59:08] dbug: 16:59:08.5182020 Xamarin.Hosting: No need to boot (already booted): iPhone X (iOS 13.5) - created by XHarness
[16:59:08] dbug: 16:59:08.5218370 Xamarin.Hosting: Installing /tmp/helix/working/9F630914/w/A3C008F7/e/Microsoft.Extensions.Configuration.Functional.Tests.app with Bundle Identifier net.dot.Microsoft.Extensions.Configuration.Functional.Tests on 'iOS 13.5 (17F61) - iPhone X'...
[17:28:44] dbug: 17:28:44.9031320 Killing process 14971 as it was cancelled
[17:28:44] dbug: 17:28:44.9088890 Process exited with 137
[17:28:44] dbug: 17:28:44.9095210 Test run failed
[17:28:44] dbug: 17:28:44.9175530 Could not find pid in mtouch output.
[17:28:44] dbug: 17:28:44.9224100 Test run started
[17:28:46] dbug: 17:28:46.0798530 Test run failed to launch
[17:28:46] dbug: 17:28:46.0844090 No crash reports, waiting a second to see if the crash report service just didn't complete in time (4)
[17:28:47] dbug: 17:28:47.2295520 No crash reports, waiting a second to see if the crash report service just didn't complete in time (3)
[17:28:48] dbug: 17:28:48.3149390 No crash reports, waiting a second to see if the crash report service just didn't complete in time (2)
[17:28:49] dbug: 17:28:49.4627860 No crash reports, waiting a second to see if the crash report service just didn't complete in time (1)
[17:28:50] dbug: 17:28:50.5966390 No crash reports, waiting a second to see if the crash report service just didn't complete in time (0)
[17:28:51] fail: Failed to launch the application:
                 Launch failure
                 
                 Check logs for more information.
XHarness exit code: 83
+ exit_code=83
```

https://github.com/dotnet/runtime/pull/45172
https://dev.azure.com/dnceng/public/_build/results?buildId=898906&view=results
https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-45172-merge-f3086f428107446aa7/Microsoft.Extensions.Configuration.Functional.Tests/console.c96502f5.log?sv=2019-07-07&se=2020-12-15T00%3A55%3A49Z&sr=c&sp=rl&sig=i1uWKdwS6iGhcUW0XoWG5uWfxxQF%2FwXqK6V6beKh6YM%3D

cc: @premun @dotnet/runtime-infrastructure

**Comment by premun** [comment]
Having a look. This happened before when the MacOS sometimes slows down and I was waiting for this to prepare a final fix.

Usually it's tied to a specific machine, so I will take them offline to unblock runtime and prepare the fix.

**Comment by premun** [comment]
@safern I took these machines offline:
- `dci-mac-build-133`
- `dci-mac-build-136`
- `dci-mac-build-146`

The rest looked healthy from a quick Kusto search.
I will add work item retry + machine reboot for these cases and that should resolve this for good.

**Comment by radical** [comment]
Another instance of this, IIUC, on https://github.com/dotnet/runtime/pull/45839:

https://dnceng.visualstudio.com/public/_build/results?buildId=914235&view=logs&j=8580ecfb-912a-5dbd-35ce-e64d0d51ddbf&t=ebd4788e-80c8-5104-29f3-a1792da138c8

console: https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-45839-merge-13b48d1686ba4343b7/Microsoft.Bcl.AsyncInterfaces.Tests/console.fa2f6cc5.log?sv=2019-07-07&se=2020-12-29T16%3A46%3A22Z&sr=c&sp=rl&sig=tn4cgwt0brU45SKlzBlHEbsrbDNIhpZNsqrQkidt7Ss%3D

This is for `Microsoft.Bcl.AsyncInterfaces.Tests`. Also, I see an extra message before the process gets killed (2nd line):
```
[09:16:32] dbug: 09:16:32.4000240 Xamarin.Hosting: Installed 'net.dot.Microsoft.Bcl.AsyncInterfaces.Tests' from /tmp/helix/working/A1B108B5/w/B28E0A07/e/Microsoft.Bcl.AsyncInterfaces.Tests.app
[09:16:39] dbug: 09:16:39.9175400 Xamarin.Hosting: Can't process potentially bad load commands, because the app is signed.
[09:16:52] dbug: 09:16:52.6508610 Killing process 943 as it was cancelled
```

**Comment by radical** [comment]
And two more library test runs fail with the same issue, on the PR #45839 from the last comment:
[System.Data.Odbc.Tests](https://helix.dot.net/api/2019-06-17/jobs/13b48d16-86ba-4343-b79e-5191586b11d2/workitems/System.Data.Odbc.Tests/console), and [System.Runtime.Serialization.Json.ReflectionOnly.Tests](https://helix.dot.net/api/2019-06-17/jobs/13b48d16-86ba-4343-b79e-5191586b11d2/workitems/System.Data.Odbc.Tests/console).

**Comment by steveisok** [comment]
dci-mac-build-133, dci-mac-build-145, and dci-mac-build-068.local are the 3 machines that have any badexits in the last 24 hours.

**Comment by premun** [comment]
We have a fix in Arcade that was supposed to retry the test on another machine in this case and reboot the machine.
However, the fix seems to have failed:
```
+ /usr/local/bin/python3 -c 'from helix.workitemutil import request_infra_retry; request_infra_retry('\''Retrying because iOS Simulator application install hung'\'')'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'helix'
+ /usr/local/bin/python3 -c 'from helix.workitemutil import request_reboot; request_reboot('\''Rebooting because iOS Simulator application install hung '\'')'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'helix'
```

I disabled the machines and will have a look at the retry

**Comment by ViktorHofer** [comment]
Happened again in https://dev.azure.com/dnceng/public/_build/results?buildId=915078&view=ms.vss-test-web.build-test-results-tab&runId=29120420&resultId=142002&paneView=attachments

**Comment by ViktorHofer** [comment]
Reopening until we get the change into dotnet/runtime.

**Comment by premun** [comment]
We have an option - once this finishes: https://dev.azure.com/dnceng/public/_build/results?buildId=915958&view=results
we can open a draft PR in runtime, bump Arcade to this newer version there (not merge it), and let the fix go round the machines and reboot them once the iOS legs starts.

**Comment by premun** [comment]
@ViktorHofer these machines have been taken out of rotation 9 hours ago already, the builds should go green before we deploy the fix.

**Comment by ViktorHofer** [comment]
Sounds good. Let's use this issue then to track getting the fix into dotnet/runtime.

**Comment by premun** [comment]
I will open the draft PR tomorrow and return the bad machines to the pool at the same time so that I am in control more. Now there's too much traffic and I will be off soon anyway

**Comment by safern** [comment]
I will be moving this builds to the staging pipeline and then we can evaluate the fix there for some days.

**Comment by ViktorHofer** [comment]
The package is already published from the official build and we could already open a PR. Shouldn't we wait for this before moving the builds out?

**Comment by ViktorHofer** [comment]
OK the package isn't out yet.

**Comment by premun** [comment]
I didn't have time for this today as other blocker appeared but since we are in staging pipeline, I think we're fine with this on Monday too?

**Comment by steveisok** [comment]
Yes, Monday is fine.  That's why it made sense to move it over to the staging pipeline.  Gives us a bit of flex.

**Comment by premun** [comment]
Seems that runtime got the Arcade version needed for this organically already. I put the machines back into the pool and am monitoring the builds in the staging pipeline.

**Comment by premun** [comment]
All seems OK. The iOS leg has been passing steadily for last few days, just last weekend was red thanks to the new machines that were added and were misconfigured.

I think we can consider adding iOS back to non-staging

**Comment by steveisok** [comment]
@premun Do you log something when reboots occur?  I'd like to validate when that happens before moving off the staging pipeline.

**Comment by premun** [comment]
We know when the last reboot happened only. This seems to be Dec 9th for most of the machines.

However, I observed the issue again here and the fix of the fix is not working still:
https://dev.azure.com/dnceng/public/_build/results?buildId=920110&view=logs&j=8580ecfb-912a-5dbd-35ce-e64d0d51ddbf&t=ebd4788e-80c8-5104-29f3-a1792da138c8

```
+ '[' 83 -eq 83 ']'
+ export PYTHON_PATH=::/etc/helix/scripts
+ PYTHON_PATH=::/etc/helix/scripts
+ /usr/local/bin/python3 -c 'from helix.workitemutil import request_infra_retry; request_infra_retry('\''Retrying because iOS Simulator application install hung'\'')'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'helix'
+ /usr/local/bin/python3 -c 'from helix.workitemutil import request_reboot; request_reboot('\''Rebooting because iOS Simulator application install hung '\'')'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'helix'
```

So we're still not quite there yet

**Comment by premun** [comment]
The problematic machines are still 133 and 145, like before. Last reboot on 9th

**Comment by premun** [comment]
I prepared a fix and tomorrow morning when there's no traffic on the queue, I will bump Arcade in a draft PR and run the iOS leg again to see.

**Comment by safern** [comment]
I removed `blocking-clean-ci` since this leg was moved the staging pipeline.

**Comment by ViktorHofer** [comment]
We probably need a new label for the staging pipeline similar to the outerloop one: blocking-clean-ci-staging.

**Comment by safern** [comment]
I would suggest something like `staging` or something like that. As the staging pipeline is not `blocking` per se since that is it's purpose. To test things that are not stable before moving them to the main pipeline.

**Comment by ViktorHofer** [comment]
Sounds good, just a small addition, the staging pipeline can be blocking if there are build failures on it. We just hit that in another PR.

**Comment by ViktorHofer** [comment]
Also I think the `blocking` refers to the `clean` and the pipeline that follows afterwards: `blocking-clean-ci-staging`. I read this is as, this issue is blocking a clean ci on the staging pipeline.

**Comment by safern** [comment]
Fair point. Yeah we need to fail on build failures to protect verticals for new platforms that are part of the staging pipeline.

**Comment by premun** [comment]
So I did some investigations and the problem is that we are using `launchctl asuser` to start XHarness in a user session with GUI rendering capabilities in order to be able to run the Simulator.
However, this process doesn't run in a subshell so we lose all the env variables and that's why we cannot request an infra retry. I will need to take care of this outside of `launchctl` so will need more time on this.

**Comment by premun** [comment]
I returned the bad machines to the pool and ran a new job with the latest fix. However, yesterday's rollout rebooted the machines and I couldn't observe the fix anymore since they're not broken anymore. We will have to wait for another machine going bad.. 🙄

@steveisok there won't be any rollout until January so we might get another shot. Do we keep this open until then or re-open if we see the issue again?

**Comment by akoeplinger** [comment]
I'd keep this open until we can confirm it helped, if it's closed we'll forget to check :)

**Comment by premun** [comment]
I did a quick search and found following:
1. There was [a build](https://dev.azure.com/dnceng/public/_build/results?buildId=930299&view=logs&j=8580ecfb-912a-5dbd-35ce-e64d0d51ddbf&t=9dfcea74-7771-586a-c822-c74db2a2a635&l=49) where one machine's Simulator started to slow down

2. We failed to install the app during [one of the work items (System.Xml.Linq.SDMSample.Tests) which timed out after 30 minutes](https://helix.dot.net/api/jobs/4f3d48c1-3e77-4e43-94ae-b6089755436c/workitems/System.Xml.Linq.SDMSample.Tests?api-version=2019-06-17) (this is what we set for XHarness), which unfortunately took just few minutes too long for the whole build to finish in time and the build got cancelled

3. The fix from Arcade worked correctly and re-tried the work item:
```
+ /usr/local/bin/python3 -c 'from helix.workitemutil import request_infra_retry; request_infra_retry('\''Retrying because iOS Simulator application install hung'\'')'
2020-12-23T09:16:18.409Z	INFO   	console_command	workitemutil(23)	request_infra_retry	Infrastructure retry (rerun of work item) has been requested
2020-12-23T09:16:18.409Z	INFO   	console_command	workitemutil(48)	_write_workitem_status_file	Created file 'infraretry-e9f4bdd0-108a-49c8-8cc5-252a78335920-4f3d48c1-3e77-4e43-94ae-b6089755436c.txt' to indicate a request for infraretry
+ /usr/local/bin/python3 -c 'from helix.workitemutil import request_reboot; request_reboot('\''Rebooting because iOS Simulator application install hung '\'')'
2020-12-23T09:16:18.893Z	INFO   	console_command	workitemutil(14)	request_reboot	Post-workitem reboot has been requested.
2020-12-23T09:16:18.894Z	INFO   	console_command	workitemutil(48)	_write_workitem_status_file	Created file 'reboot-e9f4bdd0-108a-49c8-8cc5-252a78335920-4f3d48c1-3e77-4e43-94ae-b6089755436c.txt' to indicate a request for reboot
```

4. The re-tried work item succeeded and the whole job succeeded
![image](https://user-images.githubusercontent.com/7013027/102990209-50f91e80-4517-11eb-83ef-8262a95db8a2.png)

5. The machine which was rebooted **dci-mac-build-068.local** has been succeeding since then:
![image](https://user-images.githubusercontent.com/7013027/102989438-e85d7200-4515-11eb-8bcd-830f6f36ab8d.png)


So...

If we lower the `LaunchTimeout` which cares about the time between **XHarness started** and **iOS application is booted and pings XHarness for the first time**, we could improve build times where the app install is hanging and we won't potentially even have to increase the timeout for the iOS leg. However, sometimes the build part of the leg takes even 1:45:00 which only gives 15 minutes to the test stage.
Usually, iOS tests take 6-9 minutes but if the queue gets more full, this can also tip the build over already.
I would:

1. Lower the `LaunchTimeout` of the XHarness work items to 10-15 minutes
2. Increase the build leg timeout to 2:30:00

What do you say?

/cc @akoeplinger @steveisok

**Comment by akoeplinger** [comment]
@premun Yes, that sounds good, we should do both of the steps you suggest.

**Comment by steveisok** [comment]
Fine by me.  Thanks for validating.

**Comment by safern** [comment]
> However, sometimes the build part of the leg takes even 1:45:00 

Is there something we can do to improve this? The test build is quite long on iOS, right?

**Comment by premun** [comment]
We are building one app at a time so I can imagine we could parallelize this but I guess a question for @akoeplinger

**Comment by akoeplinger** [comment]
Building the tests in parallel is one option but to get build time really down we need to pack multiple test assemblies in one or at least a few iOS app(s).

We've talked about this in the past but I'm not 100% sure how much work it'd be in the libraries tests build, we'd need to have a step that runs after the normal tests build that can take a subset of the tests and package them into an iOS app and we'd pick the entrypoint assembly when we start the app via xharness like we already do for the runtime tests.

**Comment by safern** [comment]
reopened as we need to track this getting back from runtime-staging into the main pipeline.

**Comment by premun** [comment]
Fix for the exit code detection was merged yesterday and runtime has the new version since this morning.

Let's give it a couple of days but iOS leg should be stable now, so let's see.

**Comment by premun** [comment]
Looks like if the issue for which we're doing retry+reboot happens for `ios run` instead of `ios test` together with `--expecited-exit-code`, we don't get back the right return code and the retry+reboot won't happen. So this needs to be fixed as it happened at least once since the latest fix.

**Comment by ViktorHofer** [comment]
@premun what's the current status on this? Unsure what the remaining work this issue is tracking for xHarness.

**Comment by premun** [comment]
@ViktorHofer the pipeline has been green for a long time then it wasn't but not because of infrastructure issues but because of actual failing tests. I believe those were fixed and moving the pipeline back to the official build was just the matter if we want to and it would bring us value at the moment or we don't care about the platform enough yet. But infrastracture-wise, I don't know about any issues now.

Moving it back is totally up to @steveisok and we discuss it on our regular syncs. However, I am not monitoring it at the moment because I don't believe we reached a decision to move it back.

**Comment by ViktorHofer** [comment]
> But infrastracture-wise, I don't know about any issues now.

Sounds to me like we can close the issue then. 

> Moving it back is totally up to @steveisok and we discuss it on our regular syncs. However, I am not monitoring it at the moment because I don't believe we reached a decision to move it back.

@steveisok would it make sense to open new issues to track the platforms that should run as part of the main runtime pipeline instead of in runtime-staging?

**Comment by steveisok** [comment]
> 
> @steveisok would it make sense to open new issues to track the platforms that should run as part of the main runtime pipeline instead of in runtime-staging?

Yes, however, they are on hold to move to the main runtime until we have more resiliency options.
